File: LiveCapture/LiveCapture/Core/Intelligence/SceneModels.swift
//
//  SceneModels.swift
//  LiveCapture
//
//  AI 场景智能引擎 - 数据模型定义
//
//  ## 文件作用
//  定义场景识别、环境光分析、主体检测和自适应拍摄参数的数据模型
//  所有模型支持 Codable，便于持久化和调试
//
//  ## 主要类型
//
//  ### SceneType 枚举
//  支持 14+ 场景类型识别，涵盖日常拍摄常见场景
//  - portrait: 人像
//  - food: 美食
//  - landscape: 风景
//  - pet: 宠物
//  - architecture: 建筑
//  - nightScene: 夜景
//  - document: 文档
//  - sunrise: 日出日落
//  - snow: 雪景
//  - beach: 海滩
//  - flower: 花卉
//  - stage: 舞台
//  - street: 街拍
//  - indoor: 室内
//  - unknown: 未知
//
//  ### LightAnalysis 结构体
//  环境光分析结果，包含色温、亮度、对比度、逆光标和光源类型
//
//  ### SubjectDetection 结构体
//  主体检测结果，包含人物、动物、食物检测信息
//
//  ### AdaptiveCaptureParams 结构体
//  场景自适应拍摄参数，包含 ISO、快门、曝光补偿、白平衡、变焦建议等
//

import Foundation
import CoreGraphics

#if os(iOS)

/// 场景类型枚举 - 支持 14+ 场景识别
enum SceneType: String, CaseIterable, Codable {
	case portrait      // 人像
	case food          // 美食
	case landscape     // 风景
	case pet           // 宠物
	case architecture  // 建筑
	case nightScene    // 夜景
	case document      // 文档
	case sunrise       // 日出日落
	case snow          // 雪景
	case beach         // 海滩
	case flower        // 花卉
	case stage         // 舞台
	case street        // 街拍
	case indoor        // 室内
	case unknown       // 未知

	/// 场景中文显示名称
	var displayName: String {
		switch self {
		case .portrait:      return "人像"
		case .food:          return "美食"
		case .landscape:     return "风景"
		case .pet:           return "宠物"
		case .architecture:  return "建筑"
		case .nightScene:    return "夜景"
		case .document:      return "文档"
		case .sunrise:       return "日出日落"
		case .snow:          return "雪景"
		case .beach:         return "海滩"
		case .flower:        return "花卉"
		case .stage:         return "舞台"
		case .street:        return "街拍"
		case .indoor:        return "室内"
		case .unknown:       return "自动"
		}
	}

	/// SF Symbol 图标名称
	var iconName: String {
		switch self {
		case .portrait:      return "person.crop.rectangle"
		case .food:          return "fork.knife"
		case .landscape:     return "mountain.2"
		case .pet:           return "pawprint"
		case .architecture:  return "building.2"
		case .nightScene:    return "moon.stars"
		case .document:      return "doc.text"
		case .sunrise:       return "sunrise"
		case .snow:          return "snowflake"
		case .beach:         return "beach.umbrella"
		case .flower:        return "flower"
		case .stage:         return "theatermasks"
		case .street:        return "road.lanes"
		case .indoor:        return "house"
		case .unknown:       return "camera"
		}
	}
}

/// 环境光分析结果
struct LightAnalysis {
	/// 色温 (K)，范围约 2000-10000
	let colorTemperature: Float
	/// 亮度 (0-1)，0 为全黑，1 为过曝
	let brightness: Float
	/// 对比度 (0-1)，0 为完全平坦，1 为极高对比
	let contrast: Float
	/// 是否逆光场景
	let isBacklit: Bool
	/// 光源类型
	let lightType: LightType

	/// 光源类型枚举
	enum LightType: String, Codable {
		case natural      // 自然光
		case warm        // 暖光
		case cool        // 冷光
		case fluorescent // 荧光灯
		case mixed       // 混合光源
	}

	/// 默认环境光分析结果（中性值）
	static let `default` = LightAnalysis(
		colorTemperature: 5500,
		brightness: 0.5,
		contrast: 0.3,
		isBacklit: false,
		lightType: .natural
	)
}

/// 主体检测结果
struct SubjectDetection {
	/// 是否检测到人物
	let hasHuman: Bool
	/// 是否检测到动物
	let hasAnimal: Bool
	/// 是否检测到食物
	let hasFood: Bool
	/// 检测到的人物数量
	let humanCount: Int
	/// 主要主体在画面中的归一化矩形 [0,1]
	let mainSubjectRect: CGRect
	/// 主体类型描述
	let subjectType: String

	/// 默认主体检测结果
	static let `default` = SubjectDetection(
		hasHuman: false,
		hasAnimal: false,
		hasFood: false,
		humanCount: 0,
		mainSubjectRect: CGRect(x: 0.25, y: 0.25, width: 0.5, height: 0.5),
		subjectType: "未知"
	)
}

/// 场景自适应拍摄参数
struct AdaptiveCaptureParams: Codable {
	/// 目标 ISO 值
	let targetISO: Float
	/// 目标快门速度（秒）
	let targetShutterSpeed: Float
	/// 曝光补偿 (EV)
	let exposureBias: Float
	/// 白平衡色调偏移 (-1 到 1)
	let whiteBalanceTint: Float
	/// 白平衡色温 (K)
	let whiteBalanceTemperature: Float
	/// 建议变焦倍数
	let suggestedZoomFactor: CGFloat
	/// 建议镜头类型
	let suggestedLensType: String
	/// 闪光灯建议
	let flashMode: FlashRecommendation

	/// 闪光灯推荐模式
	enum FlashRecommendation: String, Codable {
		case auto  // 自动
		case on    // 开启
		case off   // 关闭
	}

	/// 默认拍摄参数
	static let defaultParams = AdaptiveCaptureParams(
		targetISO: 200,
		targetShutterSpeed: 1.0 / 120.0,
		exposureBias: 0.0,
		whiteBalanceTint: 0.0,
		whiteBalanceTemperature: 5500,
		suggestedZoomFactor: 1.0,
		suggestedLensType: "wide",
		flashMode: .auto
	)
}

#endif

File: LiveCaptureAndroid/app/src/main/java/com/livecompose/livecapture/core/intelligence/SceneModels.kt
package com.livecompose.livecapture.core.intelligence

import android.graphics.RectF

// =============================================================================
// 1. SceneType — 场景类型枚举
// =============================================================================

enum class SceneType(val chineseName: String) {
    PORTRAIT("人像"),
    PORTRAIT_STANDING("站立人像"),
    PORTRAIT_SITTING("坐姿人像"),
    COUPLE("情侣"),
    CHILDREN("儿童"),
    FOOD("美食"),
    LANDSCAPE("风景"),
    WEDDING("婚礼"),
    PRODUCT("产品"),
    BACKLIT("逆光"),
    UNKNOWN("未知"),
    GROUP("合影"),
    SELFIE("自拍"),
    NIGHT("夜景"),
    SUNSET("日落"),
    URBAN("城市"),
    NATURE("自然"),
    INDOOR("室内"),
    OUTDOOR("室外"),
    SILHOUETTE("剪影"),
    ACTION("动作"),
    STILL_LIFE("静物"),
    MACRO("微距"),
    TRAVEL("旅行"),
    STREET("街拍"),
    ARCHITECTURE("建筑"),
    PET("宠物"),
    SPORTS("运动"),
    EVENT("活动"),
    FASHION("时尚"),
    BEAUTY("美妆"),
    DOCUMENTARY("纪实"),
    MINIMAL("极简"),
    VINTAGE("复古"),
    CINEMATIC("电影感")
}

// =============================================================================
// 2. PoseCategory — 姿势分类
// =============================================================================

enum class PoseCategory {
    PORTRAIT_STANDING,
    PORTRAIT_SITTING,
    COUPLE,
    CHILDREN,
    FOOD,
    LANDSCAPE,
    WEDDING,
    PRODUCT
}

// =============================================================================
// 3. PoseDifficulty — 姿势难度
// =============================================================================

enum class PoseDifficulty {
    BEGINNER, INTERMEDIATE, ADVANCED
}

// =============================================================================
// 4. PoseGender — 姿势性别适配
// =============================================================================

enum class PoseGender {
    MALE, FEMALE, ANY
}

// =============================================================================
// 5. PoseKeypoints — 姿势关键点
// =============================================================================

data class PoseKeypoints(
    val shoulders: String,
    val head: String,
    val arms: String,
    val legs: String,
    val hands: String,
    val back: String,
    val eyeContact: String,
    val bodyAngle: String,
    val hips: String,
    val feet: String,
    val chin: String,
    val gaze: String
)

// =============================================================================
// 6. PoseTemplate — 姿势模板
// =============================================================================

data class PoseTemplate(
    val id: String,
    val name: String,
    val category: PoseCategory,
    val keypoints: PoseKeypoints,
    val tips: List<String>,
    val variations: List<String>,
    val difficulty: PoseDifficulty = PoseDifficulty.INTERMEDIATE,
    val gender: PoseGender = PoseGender.ANY
)

// =============================================================================
// 7. SubjectDetection — 被摄主体检测结果
// =============================================================================

data class SubjectDetection(
    val faceCount: Int = 0,
    val faceRect: FaceRect? = null,
    val bodyRect: BodyRect? = null,
    val faceConfidence: Float = 0.0f,
    val hasMultiplePeople: Boolean = false,
    val dominantSubjectPosition: SubjectPosition = SubjectPosition.CENTER,
    val subjectSizeRatio: Float = 0.0f,
    val detectedGenders: List<PoseGender> = emptyList(),
    val detectedAges: List<AgeGroup> = emptyList()
)

data class FaceRect(
    val left: Float,
    val top: Float,
    val right: Float,
    val bottom: Float,
    val confidence: Float = 1.0f
) {
    val centerX: Float get() = (left + right) / 2f
    val centerY: Float get() = (top + bottom) / 2f
    val width: Float get() = right - left
    val height: Float get() = bottom - top
    val area: Float get() = width * height
}

data class BodyRect(
    val left: Float,
    val top: Float,
    val right: Float,
    val bottom: Float
)

enum class SubjectPosition {
    CENTER, TOP_LEFT, TOP_CENTER, TOP_RIGHT,
    MIDDLE_LEFT, MIDDLE_RIGHT,
    BOTTOM_LEFT, BOTTOM_CENTER, BOTTOM_RIGHT
}

enum class AgeGroup {
    CHILD, TEEN, YOUNG_ADULT, ADULT, SENIOR
}

// =============================================================================
// 8. PoseSuggestion — 姿势建议
// =============================================================================

data class PoseSuggestion(
    val id: String,
    val title: String,
    val description: String,
    val instructions: String,
    val tips: List<String>,
    val assetPath: String,
    val priority: Float,
    val category: PoseCategory,
    val difficulty: PoseDifficulty = PoseDifficulty.INTERMEDIATE,
    val isAdjustment: Boolean = false,
    val isComposition: Boolean = false,
    val isDynamic: Boolean = false
)

// =============================================================================
// 9. PoseRecommendationResult — 姿势推荐结果
// =============================================================================

data class PoseRecommendationResult(
    val suggestions: List<PoseSuggestion>,
    val primaryRecommendation: PoseSuggestion?,
    val sceneType: SceneType,
    val confidenceScore: Float,
    val generatedAt: Long = System.currentTimeMillis()
) {
    val hasRecommendations: Boolean
        get() = primaryRecommendation != null || suggestions.isNotEmpty()

    companion object {
        val EMPTY = PoseRecommendationResult(
            suggestions = emptyList(),
            primaryRecommendation = null,
            sceneType = SceneType.UNKNOWN,
            confidenceScore = 0.0f
        )
    }
}

// =============================================================================
// 10. LightAnalysis — 光线分析
// =============================================================================

enum class LightType {
    NATURAL, WARM, COOL, FLUORESCENT, MIXED
}

data class LightAnalysis(
    val colorTemperature: Float,
    val brightness: Float,
    val contrast: Float,
    val isBacklit: Boolean,
    val lightType: LightType
) {
    companion object {
        val DEFAULT = LightAnalysis(
            colorTemperature = 5500f,
            brightness = 0.5f,
            contrast = 0.5f,
            isBacklit = false,
            lightType = LightType.NATURAL
        )
    }
}

// =============================================================================
// 11. AdaptiveCaptureParams — 自适应拍摄参数
// =============================================================================

enum class FlashRecommendation {
    AUTO, ON, OFF
}

data class AdaptiveCaptureParams(
    val targetISO: Float,
    val targetShutterSpeed: Float,
    val exposureBias: Float,
    val whiteBalanceTint: Float,
    val whiteBalanceTemperature: Float,
    val suggestedZoomFactor: Float,
    val suggestedLensType: String,
    val flashMode: FlashRecommendation
) {
    companion object {
        val DEFAULT = AdaptiveCaptureParams(
            targetISO = 100f,
            targetShutterSpeed = 1f / 60f,
            exposureBias = 0f,
            whiteBalanceTint = 0f,
            whiteBalanceTemperature = 5500f,
            suggestedZoomFactor = 1.0f,
            suggestedLensType = "wide",
            flashMode = FlashRecommendation.AUTO
        )
    }
}

// =============================================================================
// 12. QualityGrade / QualityAssessment — 画质评估
// =============================================================================

enum class QualityGrade(val displayName: String) {
    EXCELLENT("极佳"),
    GOOD("良好"),
    FAIR("一般"),
    POOR("较差")
}

data class QualityAssessment(
    val overallScore: Float,
    val sharpnessScore: Float,
    val noiseLevel: Float,
    val exposureScore: Float,
    val colorHarmonyScore: Float,
    val resolutionScore: Float,
    val qualityGrade: QualityGrade,
    val timestamp: Long,
    val imageInfo: ImageInfo?
)

data class ImageInfo(
    val width: Int,
    val height: Int,
    val aspectRatio: Float,
    val orientation: String,
    val format: String,
    val totalPixels: Long,
    val resolutionLevel: String
)

// =============================================================================
// 13. ColorAnalysis — 色彩分析
// =============================================================================

data class AverageColor(
    val r: Int, val g: Int, val b: Int, val hex: String
)

data class DominantColor(
    val r: Int, val g: Int, val b: Int, val hex: String, val percentage: Float
)

data class ColorTemperatureInfo(
    val kelvin: Float, val type: String, val description: String
)

data class ColorAnalysis(
    val averageColor: AverageColor,
    val dominantColors: List<DominantColor>,
    val colorTemperature: ColorTemperatureInfo,
    val colorHarmonyScore: Float,
    val colorMood: String,
    val saturationMean: Float,
    val brightnessMean: Float,
    val contrastRatio: Float
)

// =============================================================================
// 14. EnhancementType / EnhancementSuggestion — 增强建议
// =============================================================================

enum class EnhancementType {
    SHARPNESS, NOISE_REDUCTION, EXPOSURE, COLOR_HARMONY,
    PORTRAIT_ENHANCEMENT, LANDSCAPE_ENHANCEMENT,
    NIGHT_OPTIMIZATION, FOOD_ENHANCEMENT
}

enum class Priority {
    HIGH, MEDIUM, LOW
}

data class EnhancementSuggestion(
    val type: EnhancementType,
    val title: String,
    val description: String,
    val parameters: Map<String, Float>,
    val priority: Priority
)

// =============================================================================
// 15. ScenePresetParams — 场景预设参数
// =============================================================================

data class ScenePresetParams(
    val exposure: Float = 0f,
    val contrast: Float = 0f,
    val saturation: Float = 0f,
    val highlights: Float = 0f,
    val shadows: Float = 0f,
    val clarity: Float = 0f,
    val warmth: Float = 0f,
    val sharpness: Float = 0f,
    val noiseReduction: Float = 0f,
    val vignette: Float = 0f
) {
    companion object {
        val DEFAULT = ScenePresetParams()
    }
}

// =============================================================================
// 16. CompositionAnalysis — 构图分析
// =============================================================================

data class CompositionAnalysis(
    val ruleOfThirdsScore: Float,
    val symmetryScore: Float,
    val visualBalanceScore: Float,
    val leadingLinesCount: Int,
    val focalPointsCount: Int,
    val compositionType: String,
    val feedback: String
)